### ISSUE TITLE
`no-unused-expressions` started to trigger on `'use strict'` in 9.25.0

### ISSUE THREAD

**Issue by zloirock** [eslint/eslint#19637]
### Environment

Node version: 23.10.0
npm version: 11.3.0
Local ESLint version: 9.25.0
Global ESLint version: -
Operating System: MacOS 15.5 Beta

### What parser are you using?

Default (Espree)

### What did you do?

`no-unused-expressions` started to trigger on `'use strict'` in 9.25.0

### What did you expect to happen?

No errors.

### What actually happened?

<img width="752" alt="Image" src="https://github.com/user-attachments/assets/36e5110d-a22d-4522-b7ba-7f55e959b8ef" />

### Additional comments

You could check it on `core-js` repo.

Config: https://github.com/zloirock/core-js/blob/master/tests/eslint/eslint.config.js
`package.json`: https://github.com/zloirock/core-js/blob/master/tests/eslint/package.json

`npm run lint` in the root.

**Comment by snitin315** [comment]
@zloirock Thanks for the report! I’m able to reproduce the issue. It seems related to a recent change in https://github.com/eslint/eslint/commit/77d6d5bc4923012aee34b0a7c3d971f017d65555.

I’m investigating the root cause and will update you soon.

**Comment by snitin315** [comment]
I found that `node.directive` is `undefined` for some unknown (yet) reason in core-js repo. And we refactored the code to rely on `node.directive` property in https://github.com/eslint/eslint/pull/19564

<img width="788" alt="Image" src="https://github.com/user-attachments/assets/8372eae0-e0c9-4347-8387-531c67dcf225" />

**Comment by snitin315** [comment]
That's correct. Here's a minimal [Playground Repro](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLWV4cHJlc3Npb25zOiBbXCJlcnJvclwiLCB7IGFsbG93U2hvcnRDaXJjdWl0OiB0cnVlLCBhbGxvd1Rlcm5hcnk6IHRydWUsIGFsbG93VGFnZ2VkVGVtcGxhdGVzOiB0cnVlIH1dKi9cblxuXCJ1c2Ugc3RyaWN0XCI7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319LCJlY21hVmVyc2lvbiI6Mywic291cmNlVHlwZSI6InNjcmlwdCJ9fX0=)

**Comment by mdjermanovic** [comment]
I think this works correctly because directives didn't exist in ES3. Note that there are already many differences in parsing, scoping, and rules regarding how directive-like statements are interpreted with ecmaVersion = 3 vs ecmaVersion > 3. For example:

```js
"use strict";

// ok with ecmaVersion: 3, parsing error "Argument name clash" with ecmaVersion: 5
function foo(a, a) {} 

```

[Playground demo with ES3](https://eslint.org/play/#eyJ0ZXh0IjoiXCJ1c2Ugc3RyaWN0XCI7XG5cbi8vIG9rIHdpdGggZWNtYVZlcnNpb246IDMsIHBhcnNpbmcgZXJyb3IgXCJBcmd1bWVudCBuYW1lIGNsYXNoXCIgd2l0aCBlY21hVmVyc2lvbjogNVxuZnVuY3Rpb24gZm9vKGEsIGEpIHt9XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX0sInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hVmVyc2lvbiI6M319fQ==)

[Playground demo with ES5](https://eslint.org/play/#eyJ0ZXh0IjoiXCJ1c2Ugc3RyaWN0XCI7XG5cbi8vIG9rIHdpdGggZWNtYVZlcnNpb246IDMsIHBhcnNpbmcgZXJyb3IgXCJBcmd1bWVudCBuYW1lIGNsYXNoXCIgd2l0aCBlY21hVmVyc2lvbjogNVxuZnVuY3Rpb24gZm9vKGEsIGEpIHt9XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX0sInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hVmVyc2lvbiI6NX19fQ==)

**Comment by mdjermanovic** [comment]
@zloirock can you provide more context on why you are using `'use strict'` in code that targets ES3?

**Comment by zloirock** [comment]
@mdjermanovic it's enougth simple to understand from the description of the project. It's JavaScript standard library polyfill, code should work everywhere, but it should work maximally correct as the execution envieronment allows - strict mode is a part of this correctness.

**Comment by zloirock** [comment]
This breaking change disallows to use this rule for `core-js` since I don't see an option how to allow `'use strict'` in this rule and setting `ecmaVersion` to another version with manual disallowing all ES3->ES5 syntax too problematic because of a lot of not so obvious things.

**Comment by mdjermanovic** [comment]
> It's JavaScript standard library polyfill, code should work everywhere, even in ES3, but it should work maximally correct as the execution environment allows - strict mode is a part of this correctness for ES5+.

In that case, the same code might work differently in different environments, because `'use strict'` is a no-op in ES3 environments, but it changes the syntax and runtime behavior in ES5+ environments. So maybe it would make sense to lint that code twice, with `ecmaVersion: 3` and `ecmaVersion: 5`, and enable this rule only for the latter? Note that, with `ecmaVersion: 3`, the parser and some other rules for which strict vs non-strict mode differences are important are already analyzing the code as if it were in non-strict mode, so they might be missing some problems that would be caught with `ecmaVersion: 5` or higher.

**Comment by zloirock** [comment]
> In that case, the same code might work differently in different environments, because 'use strict' is a no-op in ES3 environments, but it changes the syntax and runtime behavior in ES5+ environments.

Sure. Developers who write their projects only for ES5+ engines can rely on proper strict mode-related behaviour everywhere, for ES3+ can't.

> Note that, with ecmaVersion: 3, the parser and some other rules for which strict vs non-strict mode differences are important are already analyzing the code as if it were in non-strict mode, so they might be missing some problems that would be caught with ecmaVersion: 5 or higher.

This is taken into account in the codebase and is monitored by other rules and tests.

> So maybe it would make sense to lint that code twice, with ecmaVersion: 3 and ecmaVersion: 5, and enable this rule only for the latter?

Unfortunately, `'use strict'` doesn't go away for the second pass.

**Comment by fasttime** [comment]
> So maybe it would make sense to lint that code twice, with `ecmaVersion: 3` and `ecmaVersion: 5`, and enable this rule only for the latter?

This was [the recommendation](https://github.com/eslint/tsc-meetings/blob/main/notes/2022/2022-01-13.md#bug-use-strict-should-not-trigger-strict-mode-in-es3) for a previous, similar issue where users had included `"use strict"` directives in code that would run in both ES3-compilant and in new engines.

**Comment by zloirock** [comment]
I think that the best option here is to add to this rule an option to ignore directives even with `ecmaVersion: 3`.

**Comment by mdjermanovic** [comment]
We've decided to revert the directive detection change and release v9.25.1 today. That should fix this issue.

In v9.26.0, we'll reintroduce this change but add an option to ignore directives even with `ecmaVersion: 3`.
